rfc/updates.md

```diff
@@ -0,0 +1,199 @@
+
+
+
+# Constellation updates
+Things we manage for the user:
+1. VM Image
+2. Kubernetes
+3. Cilium
+4. Constellation microservices
+
+
+### VM Image
+Updating the VM Image is already implemented. 
+
+
+### Kubernetes
+
+The current manual approach is:
+
+* ssh into a control-plane node
+* curl new kubeadm version (see `versions.go`) to `/var/run/state/bin/kubeadmNew`
+* `mv /var/run/state/bin/kubeadmNew /var/run/state/bin/kubeadm && chmod +x /var/run/state/bin/kubeadm`
+* `export PATH=$PATH:/var/run/state/bin/`
+* `kubeadm upgrade plan`
+* `kubeadm upgrade apply vX.Y.Z`
+* bump K8s version in configmap/k8s-version to `vX.Y`
+* [upgrade your cluster](https://constellation-docs.edgeless.systems/constellation/workflows/upgrade) with a new VM image
+
+For more details on the first steps see the [official K8s documentation](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/).
+
+
+### Cilium
+Cilium is installed via helm. In the long term, we don't need to maintain our fork and Cilium can be updated independently using the official releases. 
+
+
+### Constellation microservices
+
+All Constellation microservices will be bundled into and therefore updated via one helm chart.
+
+
+# Automatic Updates
+
+## Extending the JoinService
+
+The CLI will use a lookup table to map the Kubernetes version from the config to URLs and hashes. Those are sent over during `constellation init` and used by the first bootstrapper. Then, the URLs and hashes are pushed to the `k8s-components-1.23.12` ConfigMap and the Kubernetes version with a reference to the `k8s-components-1.23.12` ConfigMap is pushed to `k8s-versions`.
```

Comment:
Re 2:

A single config map might create merge conflicts in kubernetes. You only want one component modifying each config map.
As we also don't want the operator to know the inner details of kubernetes component versioning, the idea is to have the CLI prepare version config maps and let it hand out a reference to that config to the operator. The  operator is then the only entity modifying the k8s-version configmap.

Re 3:

Using a unique name for components might be a good idea.

Reply:
1. Will add the explanation. 
2. We thought about that too. It is easier to separate the URLs in their own ConfigMap since we need the operator to update the k8s-version but the components list will be created by the Bootstrapper (first list) and CLI (on updates). We don't want to implement any reconcile loops there. 
3. That's a good point. I overlooked that some of the components are not released in sync with Kubernetes itself. We need to re-think some part now, since the check if we need to replace a node needs to consider more than the Kubelet version itself. Maybe the node is labeled with some form of Hash(ComponentsHashes) which then the node operator can check against? @malt3 What do you think?

Reply:
> That's a good point. I overlooked that some of the components are not released in sync with Kubernetes itself. We need to re-think some part now, since the check if we need to replace a node needs to consider more than the Kubelet version itself. Maybe the node is labeled with some form of Hash(ComponentsHashes) which then the node operator can check against? @malt3 What do you think?

Yes. Labeling a node with a hash is a great idea as it integrates well with the operator. This also make comparing the reference in the k8s-version config map to the node state easier (as we can literally compare the hashes).

Reply:
>  I believe this can be done on kubeadm join but will have to check.

https://blog.scottlowe.org/2020/11/30/assigning-node-labels-during-kubernetes-cluster-bootstrapping/
Looks like it. I'd add to the spec that we label each node with `edgeless.systems/kubernetes-components-hash:"sha256:8ae09b7e922a90fea7a4259fb096f73e9efa948ea2f09349618102a328c44b8b"` and rename the components ConfigMap to `k8s-components-1.23.12-8ae09b7e922a90fea7a4259fb096f73e9efa948ea2f09349618102a328c44b8b` or should we crop the hash or just use a UID? (will do the hash for now, can discuss in new comment chain)